judge/judgedaemon.main.php

```diff
@@ -948,244 +974,238 @@ function judge(array $row)
-            'output_diff'  => rest_encode_file($testcasedir . '/feedback/judgemessage.txt', $output_storage_limit)
-        );
-        $unsent_judging_runs[] = $new_judging_run;
-        $outstanding_data += strlen(var_export($new_judging_run, TRUE));
-
-        $now = now();
-        if (!$lastcase_correct
-            || ($now - $last_sent) >= $update_every_X_seconds
-            || $outstanding_data > $row['outputlimit'] * 1024) {
-           if (send_unsent_judging_runs($unsent_judging_runs, $row['judgingid']) === null) {
-               disable('problem', 'probid', $row['probid'], "uploading unsent judging runs failed", $row['judgingid'], (string)$row['cid']);
-               return;
-           }
-           $unsent_judging_runs = array();
-           $last_sent = $now;
-           $outstanding_data = 0;
-        }
-        logmsg(LOG_INFO, '  ' . ($result === 'correct' ? " \033[0;32m✔\033[0m" : " \033[1;31m✗\033[0m")
-            . ' ...done in ' . $runtime . 's, result: ' . $result);
-    } // end: for each testcase
-    if (!empty($unsent_judging_runs)) {
-        if (send_unsent_judging_runs($unsent_judging_runs, $row['judgingid']) === null) {
-            disable('problem', 'probid', $row['probid'], "uploading unsent judging runs failed", $row['judgingid'], (string)$row['cid']);
-            return;
-        }
+    if ($result === 'compare-error') {
+        logmsg(LOG_ERR, "comparing failed for compare script '" . $judgeTask['compare_script_id'] . "'");
+        $description = 'compare script ' . $judgeTask['compare_script_id'] . ' crashed';
+        disable('compare_script', 'compare_script_id', $judgeTask['compare_script_id'], $description, $judgeTask['judgetaskid']);
+        return false;
     }
 
-    // Sanity check: need to have had at least one testcase
-    if ($totalcases == 0) {
-        logmsg(LOG_WARNING, "No testcases judged for s$row[submitid]/j$row[judgingid]!");
+    $new_judging_run = array(
+        'runresult' => urlencode($result),
+        'runtime' => urlencode((string)$runtime),
+        'output_run'   => rest_encode_file($testcasedir . '/program.out', false),
+        'output_error' => rest_encode_file($testcasedir . '/program.err', $output_storage_limit),
+        'output_system' => rest_encode_file($testcasedir . '/system.out', $output_storage_limit),
+        'metadata' => rest_encode_file($testcasedir . '/program.meta', $output_storage_limit),
+        'output_diff'  => rest_encode_file($testcasedir . '/feedback/judgemessage.txt', $output_storage_limit),
+        'hostname' => $myhost,
+    );
+
+    if ($result === 'correct') {
+        // Post result back asynchronously. PHP is lacking multi-threading, so
+        // we just call ourselves again.
+        $judgedaemon = BINDIR . '/judgedaemon';
+        $cmd = $judgedaemon
+            . ' -e ' . $endpointID
+            . ' -t ' . $judgeTask['judgetaskid']
+            . ' -j ' . base64_encode(json_encode($new_judging_run))
```

Comment:
What's the command line length limit? If this includes program output then I could see easily blowing that up if someone increases `output_storage_limit`. I guess the fix would be to store this in a temporary file and pass that to the program instead.

For me locally `getconf ARG_MAX` is `2097152`, and if we're using base64, that works out to ~1.5MiB of data.

I also did a quick google search and see that php does have pthreads, so maybe we can do it async/with threads? No idea if it works / is worth doing though. https://stackoverflow.com/a/15501449/841300

That or we should rewrite in golang :rofl:

Reply:
:eyes: I did some google search back then and didn't find anything. Let's replace it with proper threading then.

re golang: tempting :-)

Reply:
I do see this comment linked from the pho page: https://github.com/krakjoe/pthreads/issues/929#issue-410636734

Reply:
`pthreads` seems out of question; `parallel` requires PECL IIUC, so deferring this.

To address Keith's initial comment about line length I am using now an intermediate file instead.